src/briefcase/integrations/docker.py

```diff
@@ -177,6 +177,46 @@ def verify(cls, tools: ToolCache):
         tools.docker = Docker(tools=tools)
         return tools.docker
 
+    def check_output(self, image_tag, args):
+        """Run a process inside a Docker container, capturing output.
+
+        This is a bare Docker invocation; it's really only useful for running
+        simple commands on an image, ensuring that the container is destroyed
+        afterwards. In most cases, you'll want to use an app context, rather
+        than this.
+
+        :param image_tag: The Docker image to run
+        :param args: The list of arguments to pass to the Docker instance.
+        """
+        # Any exceptions from running the process are *not* caught.
+        # This ensures that "docker.check_output()" behaves as closely to
+        # "subprocess.check_output()" as possible.
+        return self.tools.subprocess.check_output(
+            [
+                "docker",
+                "run",
+                "--rm",
+                image_tag,
+            ]
+            + args,
+        )
+
+    def prepare(self, image_tag):
+        """Ensure that the given image exists, and is cached locally.
```

Comment:
Agreed that (1) is a good idea; (2) sounds like more effort than it's worth.

Reply:
I was able to confirm that `--pull=always` does **not** always re-download the image; however, it will always initiate the `docker pull` process to confirm if a new image exists and pull it if so.

While the images we're dealing with here are (relatively) small base images, we are masking a download operation with the Wait Bar. One of the machines I test on uses a VPN that is typically throttled by CDNs.....so along with `apt`, `dnf`, and `git` operations, Docker image downloads are a bit slow. So, for people in the same bucket, the first thing they see is `Checking Docker target image fedora:37...` Wait Bar with no indication anything is really happening.

Using `subprocess.run()`:
```python
            self.tools.subprocess.run(
                ["docker", "run", "--rm", "--pull=always", image_tag, "printf", ""],
                check=True,
            )
```
I get this output for the first run:
```
[helloworld] Finalizing application configuration...
9.1: Pulling from library/rockylinux
9d28f3f24f51: Pulling fs layer
9d28f3f24f51: Verifying Checksum
9d28f3f24f51: Download complete
9d28f3f24f51: Pull complete
Digest: sha256:972a6b9cec75b2f47bf0faa425135e1bdb9945306a22f26db0b3a4ee8a289687
Status: Downloaded newer image for rockylinux:9.1
Checking Docker target image rockylinux:9.1... done
Targeting rocky:9 (Vendor base rhel)
Determining glibc version... done
Targeting glibc 2.34
Targeting Python3
```
And for subsequent runs:
```
[helloworld] Finalizing application configuration...
9.1: Pulling from library/rockylinux
Digest: sha256:972a6b9cec75b2f47bf0faa425135e1bdb9945306a22f26db0b3a4ee8a289687
Status: Image is up to date for rockylinux:9.1
Checking Docker target image rockylinux:9.1... done
Targeting rocky:9 (Vendor base rhel)
Determining glibc version... done
Targeting glibc 2.34
Targeting Python3
```

The substitution of `printf` for `echo` was to avoid the arbitrary newline. AFAIK, `printf` should be as prevalent as `echo`. They're both in [`coreutils`](https://packages.debian.org/bullseye/amd64/coreutils/filelist) anyway.

Another interesting option ;)
```python
["docker", "run", "--rm", "--pull=always", image_tag, "tail", "/dev/null"],
```

Reply:
I'm on a reasonably fast connection here at home, and I have definitely noticed the slowdown on every command after adding `--pull=always`. Given it needs to be run on *every* command, I wonder if it's actually worth it. A slightly stale Docker image isn't the worst thing in the world, especially if the overall workflow is substantially faster.

Reply:
ehh...maybe hard to say. for instance, if you pulled `ubuntu:20.04` years ago for an unrelated task, that _might_ lead to problems...

Maybe it should be limited to `create` commands. Since (as i just realized) if the Briefcase Docker image for the app already exists, pulling a new version of the base image isn't even valuable....that isn't going to result in a rebuild of the Briefcase image.

Reply:
> What sort of problem are you anticipating? An `ubuntu:20.04` image tagged 3 years ago _is_ still an ubuntu:20.04 image - there might just be some more downloads to get up to date.

I don't know; I'm mostly speaking from a extremely cautious standpoint. However, if you asked me to bet money on problems actually arising, I wouldn't.

> I'm not sure only running in `create` is entirely viable. I do fairly regular purges of my docker cache to save disk space; if I `create`, then purge, then `build`, the os-release detection code won't work.

I meant only add `--pull=always` for `create` commands; this check should always run either way.

> I think my preferred approach at this point is to drop the `--pull=always`, and accept that old Docker images _might_ be present.

I think that's fine.

Reply:
With `--pull=always` removed, we still have the issue of masking the image download with the Wait Bar.

Are you interested in using `subprocess.run()` so the user knows that Docker is downloading an image?

Reply:
Yeah - that makes sense. Change incoming.

Reply:
FYI - I also disabled the waitbar and streaming on that call, so that the native download progress bars are visible. If the image already exists, the waitbar wasn't doing much; and if the image doesn't exist, the docker progress bars are a lot more helpful than the indeterminate waitbar.